### ISSUE TITLE
Shutdown (without persist) hangs in Github actions

### ISSUE THREAD

**Issue by nihohit** [boustrophedon/pgtemp#9]
I can't repro this locally, so I don't know what the source of the issue is. The temporary DB, created with `PgTempDB::async_new().await` (so `persist` isn't set), hangs for minutes after a call to `shutdown`. There are existing connections to the DB, but according to the documentation, this shouldn't block `shutdown` when `persist` isn't set.

This happens on a runner running Ubuntu noble, with postgresql-16 installed.

**Comment by boustrophedon** [comment]
Per #10 the issue is I was dumb with how I designed the shutdown function - probably what happened was that I wrote shutdown as `fn shutdown(mut self)` but then I wanted to use it inside drop, which is implemented with `fn drop(&mut self)` so I thought "oh I'll just change the signature of shutdown". However, it's obvious in hindsight that shutdown will then be called twice.

There are a lot of options here to fix this and pretty much all of them are breaking changes (which tbh isn't that big a deal since it's not like I have a ton of users). Probably simplest solution (and also doesn't change the API) is the one you provided but give me a bit to think about it.

**Comment by nihohit** [comment]
No hurry, but the PR is a separate issue - I see in github that shutdown
never completes on the first call.

**Comment by boustrophedon** [comment]
So the PR doesn't solve the hanging issue?

**Comment by nihohit** [comment]
nope, it solves a different issue.

**Comment by boustrophedon** [comment]
Taking another look at this I wonder if the older version of pg in CI is different than what we have locally and maybe the behavior on sigkill is different, and a newer pg does better cleanup?

I can't get this to race locally with 1 or 100 connections.

**Comment by nihohit** [comment]
I don't think it's a version mismatch - checked this with CI & local versions both 16.3.
It could be an OS difference - CI is Ubuntu, local is MacOS.

**Comment by boustrophedon** [comment]
I'm also on linux with 16.3. If you run the test in release mode does it hang in CI still?

**Comment by nihohit** [comment]
Yup.

**Comment by boustrophedon** [comment]
The newest release, v0.5.0, uses fast shutdown instead of sending SIGKILL. Could you check and see if this issue still occurs with the newest version?

**Comment by jacobsvante** [comment]
Did you get this to work?

**Comment by nihohit** [comment]
No, I tried it once, the issue seemed to have persisted, but I didn't have the time to investigate.
Since then I've moved to another team, so I probably won't have another opportunity to test this, so @boustrophedon I guess the issue can be closed.

**Comment by jacobsvante** [comment]
Thanks @nihohit for the quick reply. I'm receiving the same error as you originally posted, so I don't think the issue should be closed.

@boustrophedon Have you been able to get it working in Github actions?

**Comment by boustrophedon** [comment]
@jacobsvante This is still happening on pgtemp 0.5?

**Comment by boustrophedon** [comment]
Yes, they're in the tests/ directory. I just added the sample test from this issue so we'll see what happens: https://github.com/boustrophedon/pgtemp/actions/runs/13167858114

**Comment by boustrophedon** [comment]
> running 1 test
> test pgtemp_shutdown_test ... ok

The sample provided passed in CI (and then the build failed due to a linting error in the python client tests)

**Comment by jacobsvante** [comment]
The issue was that I had not read the requirements section of the readme, sorry a bit much on my plate lately...

I had forgot to add these lines:

```yaml
      - name: Install postgres
        run: sudo apt-get install postgresql postgresql-client
      - name: Update path
        run: find /usr/lib/postgresql/ -type d -name "bin" >> $GITHUB_PATH
```

I'll see if I can find some time to add a check for this when running pgtest. Would make it less likely for stressed people like me being confused about why the tests are running for a long time and then just fail with a vague error message 😅

What is causing the issue btw?

**Comment by boustrophedon** [comment]
I don't know what would have caused your tests to hang - If I remove the postgres install and PATH lines, the tests simply error out. https://github.com/boustrophedon/pgtemp/actions/runs/13188950120/job/36817698923

**Comment by jacobsvante** [comment]
Strange. Maybe it was two tests running simultaneously. I'll see if I can find what was causing it.
